cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/CapabilityManager.java

```diff
@@ -0,0 +1,422 @@
+                                                      application.getVersion());
+      try {
+        applicationLifecycleService.removeApplication(applicationId);
+      } catch (Exception exception) {
+        LOG.error("Deleting application {} failed with exception {}", applicationId, exception);
+      }
+    }
+    capabilityStatusMap.remove(capability);
+    LOG.debug("Capability {} deleted.", capability);
+  }
+
+  private void deletePipelines(String capability) throws Exception {
+    doForAllApps(capability, this::deleteAppWithRetry);
+  }
+
+  private void stopPipelines(String capability) throws Exception {
+    doForAllApps(capability, this::stopPrograms);
+  }
+
+  private void deployAllApps(String capability, List<SystemApplication> applications) {
+    if (applications.isEmpty()) {
+      LOG.debug("Capability {} do not have apps associated with it", capability);
+      return;
+    }
+    for (SystemApplication application : applications) {
+      doWithRetry(application, this::deployApp);
+    }
+  }
+
+  private void deployApp(SystemApplication application) {
+    String version = application.getVersion() == null ? ApplicationId.DEFAULT_VERSION : application.getVersion();
+    ApplicationId applicationId = new ApplicationId(application.getNamespace(), application.getName(), version);
+    LOG.debug("Deploying app {}", applicationId);
+    try {
+      if (isAppDeployed(applicationId)) {
+        //Already deployed.
+        LOG.debug("Application {} is already deployed", applicationId);
+        return;
+      }
+      String configString = application.getConfig() == null ? null : GSON.toJson(application.getConfig());
+      applicationLifecycleService
+        .deployApp(applicationId.getParent(), applicationId.getApplication(), applicationId.getVersion(),
+                   application.getArtifact(), configString, NOOP_PROGRAM_TERMINATOR, null, null);
+    } catch (Exception ex) {
+      throw new RetryableException(ex);
+    }
+  }
+
+  private boolean isAppDeployed(ApplicationId applicationId) throws Exception {
+    try {
+      applicationLifecycleService.getAppDetail(applicationId);
+      return true;
+    } catch (ApplicationNotFoundException exception) {
+      return false;
+    }
+  }
+
+  //Find all applications for capability and call consumer for each
+  private void doForAllApps(String capability, Consumer<ApplicationId> consumer) throws Exception {
+    for (NamespaceMeta namespaceMeta : namespaceAdmin.list()) {
+      int offset = 0;
+      int limit = 100;
+      NamespaceId namespaceId = namespaceMeta.getNamespaceId();
+      EntityResult<ApplicationId> results = getApplications(namespaceId, capability, null,
+                                                            offset, limit);
+      while (results.getEntities().size() > 0) {
+        //call consumer for each entity
+        results.getEntities().forEach(consumer);
+        if (results.getEntities().size() < limit) {
```

Comment:
This check seems unnecessary. Why not just let the getApplications get called again and have it return empty set if there is no more apps?

Reply:
Removed